spec_constants/spec_constant_convolution.cpp

```diff
@@ -0,0 +1,193 @@
+  coeff_t getCoefficients() {
+    // trick the compiler a bit - problem size is always < 0, but the compiler doesn't know that
+    T val( problem_size < 0 ? problem_size : 2);
+    T val0( problem_size < 0 ? problem_size : 0);
+    return {{{val0,val,val0},{val,val,val},{val0,val,val0}}};
+  }
+
+  T getDivider() {
+    // analogous to above
+    return problem_size < 0 ? T(problem_size) : T(5);
+  }
+
+  T getInitValue() {
+    // analogous to above
+    return problem_size < 0 ? T(problem_size) : T(1);
+  }
+
+  // ids for the specialization constants
+  static constexpr s::specialization_id<coeff_t> coeff_id;
+  static constexpr s::specialization_id<T> div_id;
+
+  BenchmarkArgs args;
+
+  PrefetchedBuffer<T, 2> in_buf;
+  PrefetchedBuffer<T, 2> out_buf;
+
+  std::vector<T> in_vec;
+
+public:
+  SpecConstConvBench(const BenchmarkArgs &_args) : args(_args) {}
+
+  void setup() {
+    problem_size = (int)args.problem_size;
+
+    in_vec.resize(problem_size * problem_size);
+    std::fill(in_vec.begin(), in_vec.end(), getInitValue());
+
+    in_buf.initialize(args.device_queue, in_vec.data(), s::range<2>(problem_size, problem_size));
+    out_buf.initialize(args.device_queue, in_vec.data(), s::range<2>(problem_size, problem_size));
+  }
+
+  void run(std::vector<cl::sycl::event>& events) {
+    events.push_back(args.device_queue.submit([&](cl::sycl::handler& cgh) {
+      auto in = in_buf.template get_access<s::access::mode::read>(cgh);
+      auto out = out_buf.template get_access<s::access::mode::write>(cgh);
+
+      // set the specialization constants
+      coeff_t dynamic_coeff;
+      T dynamic_div;
+      if constexpr(AccessVariant == AccessVariants::dynamic_value) {
+        dynamic_coeff = getCoefficients();
+        dynamic_div = getDivider();
+      }
+      else if constexpr(AccessVariant == AccessVariants::spec_const_value) {
+        cgh.set_specialization_constant<coeff_id>(getCoefficients());
+        cgh.set_specialization_constant<div_id>(getDivider());
+      }
+
+      cgh.parallel_for<class ConvKernel<T, AccessVariant, InnerLoops>>(
+          in.get_range(), [=](s::item<2> item_id, s::kernel_handler h) {
+            T acc = 0;
+            coeff_t coeff;
+            T div;
+            if constexpr(AccessVariant == AccessVariants::dynamic_value) {
+              coeff = dynamic_coeff;
+              div = dynamic_div;
+            }
+            else if constexpr(AccessVariant == AccessVariants::spec_const_value) {
+              coeff = h.get_specialization_constant<coeff_id>();
+              div = h.get_specialization_constant<div_id>();
+            }
+            else if constexpr(AccessVariant == AccessVariants::constexpr_value) {
+              coeff = {{{0,2,0},{2,2,2},{0,2,0}}};
+              div = 5;
+            }
+            for(int k = 0; k < InnerLoops; ++k) {
+              for (int i = -1; i <= 1; i++) {
+                if (item_id[0] + i < 0 || item_id[0] + i >= in.get_range()[0]) continue;
```

Comment:
Doesn't `item_id[0] + i` get promoted to `size_t`? And the still check works because `0 - 1 = size_t(-1) >= in.get_range()[0]`?